src/components/GetMapTrack/GetMapTrack.tsx

```diff
@@ -1,22 +1,22 @@
-import React, { useState } from 'react'
+import React from 'react'
 import { ChallengeMap } from '../Map/ChallengeMap'
 import { UploadButton } from '../UploadButton/UploadButton'
-import { useFileContent } from '../UploadButton/useFileContent'
 import { parseGpxData } from '../../model/parser'
+import { useFiles } from '../../model/useFiles'
+import { useFilesToTextMap } from '../../model/useFileToText'
 
 const GetMapTrack = () => {
-  const [file, setFile] = useState<File | null>(null)
-  const fileContent = useFileContent(file)
+  const [files, , addFiles, clearFiles] = useFiles()
+  // file upload manipulation
+  const fileTextMap = useFilesToTextMap(files)
   return (
     <>
       <ChallengeMap
-        tracks={[
-          {
-            name: file?.name ?? 'no track selected',
-            path: parseGpxData(fileContent ?? ''),
-            color: [0, 0, 255],
-          },
-        ]}
+        tracks={[...fileTextMap.entries()].map(([file, fileText]) => ({
```

Comment:
This re-renders whenever the page re-renders. IE, It happens WAY too much. Whenever someone drags the map, it calculates this map a hundred times plus.

The fix: use a custom hook for track data.

Reply:
The problem turned out to be the dependencies list. An array or object is given a certain "ID" on conception which never changes, even if things are added to it or deleted from it. The dependency list was watching this "ID" for changes, and every time, my components were returning a brand new object or array, so it was always re-rendering. If I mutated the array or object before setting it, it would instead never re-render. So I used a property of the object, the `.size` or `.length` to watch for changes instead.